### PR TITLE
[CPDEV-93827] - add new containerd format support

### PR DESCRIPTION
**Pull request by Imadzuma**
### Description
Now we support registry configuration for containerd using `registry.configs` and `registry.mirrors` in `/etc/containerd/config.toml` (parameter `containerdConfig` from cluster.yaml):
https://github.com/Netcracker/KubeMarine/blob/0.22.0/documentation/Installation.md#cri

But this approach is deprecated from containerd v1.5.0: https://github.com/containerd/containerd/blob/v1.5.0/docs/cri/registry.md#configure-image-registry
In kubemarine we support versions v1.5.* and v1.6.*, for this reason we can fully migrate to new approach for now.

Additionally, the current approach is not compatible with ctr tool unlike the new one.

More about the new approach can be found in doc: https://github.com/containerd/containerd/blob/v1.5.1/docs/hosts.md

### Solution
Although `registry.configs` is deprecated, there is no alternative for credentials configuartion: it can be specified using `registry.configs.auth`. For this reason, we are going to support new configuration format and recommend to use it in our documents. But we don't restrict to use old format and don't provide migration patch:
* added new `containerdRegistriesConfig` field, that is used to specify registries hosts configuration format;
* added `plugins."io.containerd.grpc.v1.cri".registry.config_path` enrichment, if no old format configuration is specified;
* added exception, if specified field in both formats;
* registries enrichment is migrated to new format, if it's possible;
* added regular patch to move registries to new format;
* `containerdRegistriesConfig` is included in check_paas procedure;

Little fixes:
* Some symbols are restricted in windows/macOS OSes, but we can use those symbols in dump files. It follows not expected behavior. Those symbols were replaced with '_' symbol. See test cases for examples.
* Sometimes restore procedure fails because of failing etcd container removing using ctr: root cause in async `rm task` command, that can not be finished, when this task should be removed: added removing with force;

### How to apply
Provide steps how to apply on top previous Kubemarine version to execute on existing clusters
* [Install task/s](documentation/Installation.md#installation-tasks-description)
* Manual step 
* [Upgrade procedure](documentation/Maintenance.md#upgrade-procedure)

### Test Cases
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

**TestCase 1**: exception, if `containerdRegistriesConfig` is combined with old format fields.

Test Configuration:

- Hardware: 
- OS: 
- Inventory: `containerdRegistriesConfig` specified and `registry.mirrors` or `registry.configs.tls` specified in `containerdConfig`

Steps:

1. Run `kubemarine install`

Results:

| Before | After |
| ------ | ------ |
| JSON schema error | Exception, that specified fields can't be used together |

**TestCase 2**: exception, if `plugins."io.containerd.grpc.v1.cri".registry.config_path` is combined with old format fields.

Test Configuration:

- Hardware: 
- OS: 
- Inventory: `config_path` specified and `registry.mirrors` or `registry.configs.tls` specified in `containerdConfig`

Steps:

1. Run `kubemarine install`;
2. Run `journalctl -u containerd`;

Results:

| Before | After |
| ------ | ------ |
| Successful installation or broken images pulling (depends on which registry is used) | Exception, that specified fields can't be used together |
| errors in containerd logs, that `config_path` can't be used with `registry.mirrors` or `registry.configs.tls` | --- |

**TestCase 3**: `config_path` default enrichment 

Test Configuration:

- Hardware: 
- OS: 
- Inventory: no `config_path`, `registry.mirrors` and`registry.configs.tls` specified in `containerdConfig`. Some reristries in `containerdRegistriesConfig`.

Steps:

1. Run `kubemarine install`;

Results:

| Before | After |
| ------ | ------ |
| --- | `config_path` section added in cluster_finalyzed.yaml |
| No `/etc/containerd/certs.d` dir on nodes | `/etc/containerd/certs.d` dir on nodes. Here  registries configuration is specified in hosts.toml file in separate directory with registry name |

**TestCase 4**: `config_path` enrichment for old format

Test Configuration:

- Hardware: 
- OS: 
- Inventory: no `config_path`  specified in `containerdConfig`. `registry.mirrors` or`registry.configs.tls` specified in `containerdConfig`.
Steps:

1. Run `kubemarine install`;

Results:

| Before | After |
| ------ | ------ |
| --- | overriden `config_path` section in cluster_finalyzed.yaml |
| --- |  dir with `config_path` value name exists on nodes. Here registries configuration is specified in hosts.toml file in separate directory with registry name |

**TestCase 5**: overriden `config_path`

Test Configuration:

- Hardware: 
- OS: 
- Inventory:  `config_path`, specified in `containerdConfig`. Some registries in `containerdRegistriesConfig`.
Steps:

1. Run `kubemarine install`;

Results:

| Before | After |
| ------ | ------ |
| --- | Registries configuration is specified in old format in `/etc/containerd/config.toml` file |

**TestCase 6**: new format backup-restore

Test Configuration:

- Hardware: 
- OS: 
- Inventory:  Some registries in `containerdRegistriesConfig`.
Steps:

1. Run `kubemarine install`;
2. Run `kubemarine backup`;
3. Break `/etc/containerd/certs.d` content: remove or change some files/dirs;
4. Run `kubemarine restore`;

Results:

| Before | After |
| ------ | ------ |
| No `/etc/containerd/certs.d` content in backup archive | `/etc/containerd/certs.d` content exists in backup archive |
| `/etc/containerd/certs.d` is not restored | `/etc/containerd/certs.d` is restored |

**TestCase 7**: new format  in `check_paas`

Test Configuration:

- Hardware: 
- OS: 
- Inventory:  Some registries in `containerdRegistriesConfig`.
Steps:

1. Run `kubemarine install`;
3. Break `/etc/containerd/certs.d` content: remove or change some files/dirs;
4. Run `kubemarine check_paas`;

Results:

| Before | After |
| ------ | ------ |
| Successful check_paas | Failed `services.container_runtime.configuration` check |

**TestCase 8**: registries enrichment

Test Configuration:

- Hardware: 
- OS: 
- Inventory:  some registry specified in `registry` field.
Steps:

1. Run `kubemarine install`;

Results:

| Before | After |
| ------ | ------ |
| Registry was enriched in old format | Registry was enriched in new format |

**TestCase 9**: restricted OS symbols

Test Configuration:

- Hardware: 
- OS: Windows
- Inventory:  some restricted symbols in cluster.yaml, that will be used in dump file names. E.g. `system:basic-user` role:
```yaml
rbac:
  accounts:
  - name: basic-user
    role: system:basic-user
    namespace: kube-system
```
Steps:

1. Run `kubemarine install` with enabled dumps;

Results:

| Before | After |
| ------ | ------ |
| No real dump yaml file for `basic-user`. Empty file 'account_basic-user_system' instead of it.  | Normal `account_basic-user_system_basic-user_kube-system.yaml` dump file |

### Checklist
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Integration CI passed
- [ ] Unit tests. If Yes list of new/changed tests with brief description
- [ ] There is no merge conflicts

#### Unit tests
Indicate new or changed unit tests and what they do, if any.

### CODE REVIEW

File: kubemarine/cri/containerd.py [hunk 1]
```diff
@@ -39,9 +41,42 @@ def enrich_inventory(inventory: dict, _: KubernetesCluster) -> dict:
...
             bool(strtobool(containerd_config[runc_options_path]['SystemdCgroup']))
 
+    # Check if field for new and old configuration formats are presented
+    old_format_result, old_format_field = contains_old_format_properties(inventory)
+    new_format_result, new_format_field = contains_new_format_properties(inventory)
+
+    # Fail if fields for both formats are presented
+    if old_format_result and new_format_result:
+        raise errors.FailException(f"Invalid containerd configuration: "
+                                   f"{old_format_field} can't be set when {new_format_field} is provided")
```
Comment (theboringstuff) [inline kubemarine/cri/containerd.py:51]: Would be useful to specify in documentation that two formats could not be used simultaneously

File: kubemarine/patches/p1_reconfigure_registries.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            logger.info("No registries configured, updating containerd configuration is not needed")
+            return
+        result, _ = contains_old_format_properties(raw_inventory)
+        if result:
+            logger.info("Found old format configuration, updating containerd configuration is not needed")
+            return
+        cluster.make_group_from_roles(['control-plane', 'worker']).call(configure_containerd)
+
+
+class OverrideTargetPorts(RegularPatch):
```
Comment (theboringstuff) [inline kubemarine/patches/p1_reconfigure_registries.py:44]: Seems that the name should be different

File: kubemarine/cri/containerd.py (hunk 1 above)
Comment (Imadzuma) [reply kubemarine/cri/containerd.py:51]: added notes in documentation

File: kubemarine/patches/p1_reconfigure_registries.py (hunk 2 above)
Comment (Imadzuma) [reply kubemarine/patches/p1_reconfigure_registries.py]: fixed

File: kubemarine/cri/containerd.py [hunk 3]
```diff
@@ -212,19 +275,50 @@ def configure_crictl(group: NodeGroup) -> None:
...
+    registries_config = {}
+    for registry, host_config in cluster.inventory.get("services", {}).get("cri", {})\
+            .get('containerdRegistriesConfig', {}).items():
+        # sanitize_filepath is needed for windows/macOS, where some symbols are restricted in file_path,
+        # but they can be used in mirror name (e.g. ':')
+        registries_dir = sanitize_filepath(f'registries/{registry}', replacement_text='_')
+        os.makedirs(utils.get_dump_filepath(cluster.context, registries_dir))
+        registry_host_toml = get_config_as_toml(host_config)
+        registries_config[registry] = toml.dumps(registry_host_toml)
+        utils.dump_file(cluster, registries_config[registry], sanitize_filepath(f"{registries_dir}/hosts.toml"))
```
Comment (ilia1243) [inline kubemarine/cri/containerd.py:304]: Looks like here sanitize_filepath is redundant as utils.dump_file also does sanitizing

File: kubemarine/patches/p1_reconfigure_registries.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            logger.info("Docker cri is used, updating containerd configuration is not needed")
+            return
+        if not raw_inventory.get('registry'):
+            logger.info("No registries configured, updating containerd configuration is not needed")
+            return
+        result, _ = contains_old_format_properties(raw_inventory)
+        if result:
+            logger.info("Found old format configuration, updating containerd configuration is not needed")
+            return
+        cluster.make_group_from_roles(['control-plane', 'worker']).call(configure_containerd)
```
Comment (ilia1243) [inline kubemarine/patches/p1_reconfigure_registries.py:41]: It is necessary to `configure_ctr_flags` also

File: kubemarine/patches/p1_reconfigure_registries.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        raw_inventory = res.raw_inventory()
+        cluster = res.cluster()
+
+        if cluster.inventory['services']['cri']['containerRuntime'] == 'docker':
+            logger.info("Docker cri is used, updating containerd configuration is not needed")
+            return
+        if not raw_inventory.get('registry'):
+            logger.info("No registries configured, updating containerd configuration is not needed")
+            return
+        result, _ = contains_old_format_properties(raw_inventory)
```
Comment (ilia1243) [inline kubemarine/patches/p1_reconfigure_registries.py:37]: If does not contain `registry` and does not `contains_old_format_properties`, we enrich `config_path` and thus need to reconfigure both containerd and ctr_flags.

File: kubemarine/procedures/check_paas.py [hunk 6]
```diff
@@ -870,6 +873,17 @@ def container_runtime_configuration_check(cluster: KubernetesCluster) -> None:
                 cluster.log.debug(yaml.safe_dump(yaml.safe_load(diff.to_json())))
                 success = False
 
+            actual_registries_conf = actual_registries.get(node.get_host(), {})
+            for registry, expected_registry in expected_registries.items():
+                diff = DeepDiff(expected_registry, actual_registries_conf.get(registry))
```
Comment (ilia1243) [inline kubemarine/procedures/check_paas.py:878]: Do we need to check when some nodes have registries that are not present / enriched in cluster.yaml?

File: kubemarine/cri/containerd.py [hunk 7]
```diff
@@ -212,19 +275,50 @@ def configure_crictl(group: NodeGroup) -> None:
...
             node.put(StringIO(config_string), os_specific_associations['config_location'],
                      backup=True, sudo=True, mkdir=True)
+
+            if config_path:
+                log.debug("Uploading containerd registries configuration to %s on %s node..." %
+                          (config_path, node.get_node_name()))
+                node.sudo(f'mkdir -p {config_path}')
+                for registry, host_config in registries_config.items():
+                    node.put(StringIO(host_config), f'{config_path}/{registry}/hosts.toml',
+                             backup=True, sudo=True, mkdir=True)
```
Comment (ilia1243) [inline kubemarine/cri/containerd.py:320]: Previously we managed all the registries completely, but now only those that are present in the inventory. The question to think of if we need to delete the registries what are absent in the inventory.

File: kubemarine/core/defaults.py [hunk 8]
```diff
@@ -165,16 +166,27 @@ def apply_registry(inventory: dict, cluster: KubernetesCluster) -> dict:
...
+                containerd_config[registry_section] = {
+                    'endpoint': containerd_endpoints
+                }
+        else:
+            # Add registry info in new format
+            containerd_reg_config = inventory['services']['cri'].setdefault('containerdRegistriesConfig', {})
+            if not containerd_reg_config.get(registry_mirror_address):
+                containerd_reg_config[registry_mirror_address] = {}
+                for endpoint in containerd_endpoints:
+                    containerd_reg_config[registry_mirror_address][f'host."{endpoint}"'] = {
```
[the remark this answers is not in the document]
Comment (Imadzuma) [reply kubemarine/core/defaults.py]: After private discussion enrichment strategy was fixed: kubemarine enriches hosts from endpoints (but does not remove other hosts, specified in `containerdRegistriesConfig`) and adds `capabilities=[pull, resolve]` if it's not presented.
Added unit tests for different enrichment cases.

File: kubemarine/cri/containerd.py [hunk 9]
```diff
@@ -212,19 +275,50 @@ def configure_crictl(group: NodeGroup) -> None:
...
+    config_path = config_toml['plugins']['io.containerd.grpc.v1.cri']['registry'].get('config_path')
+
+    # Dump registries configuration
+    registries_config = {}
+    for registry, host_config in cluster.inventory.get("services", {}).get("cri", {})\
+            .get('containerdRegistriesConfig', {}).items():
+        # sanitize_filepath is needed for windows/macOS, where some symbols are restricted in file_path,
+        # but they can be used in mirror name (e.g. ':')
+        registries_dir = sanitize_filepath(f'registries/{registry}', replacement_text='_')
+        os.makedirs(utils.get_dump_filepath(cluster.context, registries_dir))
```
[the remark this answers is not in the document]
Comment (Imadzuma) [reply kubemarine/cri/containerd.py]: Added more generic approach for dump subdirs:
Added special property `create_subdir` to `dump_file` function with default value `false` (for backward compatibility). If `create_subdir=true`, kubemarine will create subdirectory from specified file path before creating file itself.

File: kubemarine/cri/containerd.py (hunk 3 above)
Comment (Imadzuma) [reply kubemarine/cri/containerd.py]: fixed

File: kubemarine/patches/p1_reconfigure_registries.py (hunk 4 above)
Comment (Imadzuma) [reply kubemarine/patches/p1_reconfigure_registries.py]: fixed

File: kubemarine/patches/p1_reconfigure_registries.py (hunk 5 above)
Comment (Imadzuma) [reply kubemarine/patches/p1_reconfigure_registries.py]: Removed `config_path` enrichment unless `containerdRegistriesConfig` is not set. As result patch is needed only for registries in new format.

File: kubemarine/procedures/check_paas.py (hunk 6 above)
Comment (Imadzuma) [reply kubemarine/procedures/check_paas.py]: I thought and decided, that yes. 
Benefits:
1. Backward compatibility with current behavior;
2. Reproducible: configuration is depended only from kubemarine specification;

Fixed `check_paas` to check both sites;
Added cleaning `config_path` directory;

File: kubemarine/cri/containerd.py (hunk 7 above)
Comment (Imadzuma) [reply kubemarine/cri/containerd.py:317]: Behavior was changed. Fore details, see comment above;

File: kubemarine/procedures/check_paas.py [hunk 10]
```diff
@@ -901,6 +904,14 @@ def container_runtime_configuration_check(cluster: KubernetesCluster) -> None:
                 cluster.log.debug(yaml.safe_dump(yaml.safe_load(diff.to_json())))
                 success = False
 
+            diff = DeepDiff(expected_registries, actual_registries.get(node.get_host(), {}))
```
Comment (ilia1243) [inline kubemarine/procedures/check_paas.py:907]: This is not aligned with the above check when we take actual config from nodes as the basis and compare it with the expected config from inventory. I think they should be aligned, but I have no strong position what should be the order.

`DeepDiff(actual_config, expected_config)` variant has semantics "what will be changed on cluster if --tasks cri.configure is called".

**Review by ilia1243 — APPROVED** [review]

File: kubemarine/procedures/check_paas.py (hunk 10 above)
Comment (Imadzuma) [reply kubemarine/procedures/check_paas.py]: fixed

**Review by ilia1243 — APPROVED** [review]

**Pull request merged**